M2M support for Third-Party Apps (#1202)
auth0/docs-v2: main/docs/get-started/applications/third-party-applications/security-controls.mdx

Third-party applications cannot be granted access to [system APIs](/docs/get-started/apis#system-apis) such as the Management API or My Account API.

## Machine-to-machine (Client Credentials)

Third-party applications support the `client_credentials` grant type for machine-to-machine access. This enables backend partner integrations and server-to-server API access without user involvement.

**Requirements and constraints:**

- The application must be a confidential client (`token_endpoint_auth_method` must not be `none`).

> - [Organizations](/docs/manage-users/organizations): Machine-to-machine access with Organizations is supported. An explicit [organization client grant](/docs/manage-users/organizations/configure-organizations/configure-organization-client-grants) is required for each organization. The `allow_any_organization` option is not permitted for third-party applications. Default client grants for third-party applications cannot be used to configure `organization_usage`.

Reviewer: this link doesn't exist. did you mean https://auth0.com/docs/manage-users/organizations/organizations-for-m2m-applications/configure-your-application-for-m2m-access?

Author: Yes

> - [Organizations](/docs/manage-users/organizations): Machine-to-machine access with Organizations is supported. An explicit [organization client grant](/docs/manage-users/organizations/configure-organizations/configure-organization-client-grants) is required for each organization. The `allow_any_organization` option is not permitted for third-party applications. Default client grants for third-party applications cannot be used to configure `organization_usage`.

Reviewer: ```suggestion
- [Organizations](/docs/manage-users/organizations): Machine-to-machine access with Organizations is supported. An explicit [organization client grant](/docs/manage-users/organizations/organizations-for-m2m-applications/configure-your-application-for-m2m-access) is required for each organization. The `allow_any_organization` option is not permitted for third-party applications. Default client grants for third-party applications cannot be used to configure `organization_usage`.
```

> - [Organizations](/docs/manage-users/organizations): Machine-to-machine access with Organizations is supported. An explicit [organization client grant](/docs/manage-users/organizations/organizations-for-m2m-applications/configure-your-application-for-m2m-access) is required for each organization. The `allow_any_organization` option is not permitted for third-party applications. Default client grants for third-party applications cannot be used to configure `organization_usage`.

Reviewer 2: This bullet point doesn't match the style of the others and seems unbalanced. Either updated this bullet point to be similar:
- An explicit [Organization client grant](/docs/manage-users/organizations/organizations-for-m2m-applications/configure-your-application-for-m2m-access) is required for each Organization.

Or update the others:
- Client-type: The application must be a confidential client

> - Not available for applications created through [Dynamic Client Registration](/docs/get-started/applications/dynamic-client-registration) or [CIMD](/docs/get-started/applications/third-party-applications/configure-third-party-applications#cimd).

Reviewer: ```suggestion
- Not available for applications created through [Dynamic Client Registration](/docs/get-started/applications/dynamic-client-registration) or [CIMD](/docs/get-started/auth0-overview/create-applications/register-applications-with-cimd).
```

**Extensibility:**

> - [Actions](/docs/customize/actions) with the credentials-exchange trigger execute normally.

Reviewer: ```suggestion
- [Actions](/docs/customize/actions) with the `credentials-exchange` trigger execute normally.
```

> - [Actions](/docs/customize/actions) with the `credentials-exchange` trigger execute normally.

Reviewer 2: Awkward to read. This falls under requirements and constraints, so suggestion:
- Configure the Auth0 Actions [`credential-exchange` trigger](/docs/customize/actions/explore-triggers/machine-to-machine-trigger/credentials-exchange-event-object)

| **Auth0 system APIs** (Management API, MFA API, My Account API, My Orgs API) | Not supported. Third-party applications cannot access system APIs in user flows. |
| **MFA during refresh token exchange** | Not supported. Refresh token transactions that trigger MFA will result in an error. |
| **Rules** | Not supported. Tenants with active Rules will receive an error when a strict third-party application triggers a login flow. |

> | **Hooks** (credentials-exchange) | Not supported. Tenants with an active credentials-exchange Hook will receive an error. Migrate to [Actions](/docs/customize/actions) for credentials-exchange extensibility. |

Reviewer: ```suggestion
| **Hooks** (credentials-exchange) | Not supported. Tenants with an active `credentials-exchange` Hook will receive an error. Migrate to [Actions](/docs/customize/actions) for `credentials-exchange` extensibility. |
```
